### PR TITLE
[#5886] feat (gvfs-fuse): Implement an in-memory file system 

### PR DESCRIPTION
**Pull request by diqiu50**
### What changes were proposed in this pull request?

Implement an in-memory filesystem for testing and validating the FUSE framework. You need to implement the PathFilesystem trait and support basic file and directory operations:

### Why are the changes needed?

Fix: #5886

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

IT

### CODE REVIEW

**Comment by FANNG1** [comment]
I think it's time to organize the code hierarchy not flatten all the code in the `src` directory.

File: clients/filesystem-fuse/src/memory_filesystem.rs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+            .map(|x| self.create_file_stat(path, x))
+            .ok_or(Errno::from(libc::ENOENT))
+    }
+
+    async fn read_dir(&self, path: &Path) -> Result<Vec<FileStat>> {
+        let file_map = self.file_map.read().unwrap();
+
+        let results: Vec<FileStat> = file_map
+            .iter()
+            .filter(|x| path_in_dir(path, x.0))
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/memory_filesystem.rs:97]: please filter out the files in sub directory

**Comment by diqiu50** [comment]
> I think it's time to organize the code hierarchy not flatten all the code in the `src` directory.

Wait until the subsequent PRs are merged before organizing. Doing the later PRs now would lead to significant conflicts.

File: clients/filesystem-fuse/src/main.rs [hunk 2]
```diff
@@ -16,69 +16,18 @@
...
-    // `HDFSFileSystem` is a file system that use `OpenDALFileSystem` to access HDFS storage.
-    //
-    // `NasFileSystem` is a filesystem that uses a locally accessible path mounted by NAS tools, such as JuiceFS.
-    //
-    // `JuiceFileSystem` is a file that use `NasFileSystem` to access JuiceFS storage.
-    //
-    // `XXXFileSystem is a filesystem that allows you to implement file access through your own extensions.
+    let _ = signal::ctrl_c().await;
+    info!("Received Ctrl+C, Unmounting gvfs...");
+    gvfs_unmount().await;
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/main.rs:30]: The mount point does't unmounted, users need to umount the directory manually.

Typical FUSE programs do not run in the foreground, and exiting with Ctrl+C is not the standard practice.
This mode is used for testing and debugging.

File: clients/filesystem-fuse/src/memory_filesystem.rs (hunk 1 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/memory_filesystem.rs:97]: fix

**Pull request merged**